src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs

```diff
@@ -177,6 +177,12 @@ protected override Conversion GetInterpolatedStringConversion(BoundExpression so
 
             if (collectionTypeKind == CollectionExpressionTypeKind.CollectionInitializer)
             {
+                if (!_binder.TryBindIEnumerableCollectionInstance(targetType, syntax, BindingDiagnosticBag.Discarded, out var collectionCreation)
+                    || collectionCreation.HasAnyErrors)
```

Comment:
`HasAnyErrors` is used for suppressing futher diagnostics, not for informing binding behavior. We should not be checking it as part of determining whether a conversion exists. For example, I would not expect the presence of any use-site diagnostics to affect whether or not a conversion exists. The return of `TryBindIEnumerableCollectionInstance` should be the only value we need to check.

Reply:
Below is a concrete example impacted by removing this check. When the check is removed, the conversion exists (which is incorrect).

If you look at the logic in `TryBindIEnumerableCollectionInstance`, it returns `true` for all `NamedTypeSymbol` and `TypeParameterSymbol` target types. But we need to detect whether we were able to successfully bind or not.

You raise a good question whether use-site diagnostics should impact the existence of the conversion. 
I don't know that we have an agreed principle to guide us here. In my opinion, if we can successfully bind the object creation, then the conversion exists; otherwise, it does not. That's my understanding from the spec and discussion with Cyrus.

```
        public void TODO2()
        {
            string source = """
                using System.Collections;
                using System.Collections.Generic;

                class NoConstructorCollection : IEnumerable<int>
                {
                    static NoConstructorCollection _c = [1, 2, 3];

                    private NoConstructorCollection() { }
                    IEnumerator IEnumerable.GetEnumerator() => null;
                    IEnumerator<int> IEnumerable<int>.GetEnumerator() => null;
                    public void Add(int i) { }
                }

                class Program
                {
                    static void F(NoConstructorCollection c) { }

                    static void Main()
                    {
                        F([]);
                    }
                }
                """;

            var comp = CreateCompilation(source);

            var tree = comp.SyntaxTrees.First();
            var model = comp.GetSemanticModel(tree);
            var collection = tree.GetRoot().DescendantNodes().OfType<CollectionExpressionSyntax>().Last();
            Assert.Equal("[]", collection.ToFullString());
            var conversion = model.GetConversion(collection);
            Assert.False(conversion.IsValid); // problem here (conversion exists if the check is removed)
        }
```

Reply:
> If you look at the logic in TryBindIEnumerableCollectionInstance, it returns true for all NamedTypeSymbol and TypeParameterSymbol target types. But we need to detect whether we were able to successfully bind or not.

Yes, I agree, this is indeed the crux of the problem: the method purports to attempt to bind as an enumerable collection and report whether that was successful, but that is not what it does.

>  In my opinion, if we can successfully bind the object creation, then the conversion exists; otherwise, it does not.

This cannot be how we determine whether the conversion exists. This has serious problems with usability, problems that we enumerated when we did interpolated strings and specifically designed it to avoid. I see that the spec says this for conversions:

* To a *struct* or *class type* `T` that implements `System.Collections.IEnumerable` where:

  * The *type* contains an *[applicable](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#11642-applicable-function-member)* instance constructor that can be invoked with no arguments.
  * For each *expression element* `Ei` there is an applicable instance or extension method `Add`  for `T` invocable with a single argument `Ei`.
  * For each *spread element* `Si` there is an applicable instance or extension method `Add` for `T` invocable with a single argument of the *iteration type* of `Si`.

This is, IMO, a spec bug. There always exists a conversion from a collection expression to a type that implements IEnumerable. Further attempting to actually construct the type can reveal further errors.

Reply:
Per offline discussion, I'm holding this PR off until we have discussion in LDM tomorrow (which may affect the design and make this PR obsolete).

Reply:
Since we didn't get to cover the spec questions in LDM, I'll move ahead with the current PR.
We're also opening a spec question for LDM about whether we should bind an object creation or do a specific lookup. I'll open up a follow-up issue on that since the PR currently deviates from the spec in that regard.